### PR TITLE
feat(bundler): add lockfile maintenance

### LINKED ISSUE #3061: Bundler lock file maintenance
Waiting on #3058

### PR DESCRIPTION
**Pull request by bilby91**
Closes #3061

Example pull request created with the new code: https://github.com/bilby91/ddsl/pull/20

### CODE REVIEW

**Review by viceice — DISMISSED** [review]

File: lib/manager/bundler/artifacts.ts
Comment (rarkins) [inline lib/manager/bundler/artifacts.ts:91]: Instead of importing `fs-extra`, could you write a new function in https://github.com/bilby91/renovate/blob/add-bundler-lock-file-maintenance/lib/util/fs.ts called `deleteLocalFile()` and use that instead?

File: lib/manager/bundler/artifacts.ts
Comment (rarkins) [inline lib/manager/bundler/artifacts.ts:100]: `bundle lock` doesn't work if you have a `Gemfile` and no `Gemfile.lock`?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/bundler/artifacts.ts
Comment (bilby91) [reply lib/manager/bundler/artifacts.ts]: Done.

File: lib/manager/bundler/artifacts.ts
Comment (bilby91) [reply lib/manager/bundler/artifacts.ts]: Seems to have done the same. I changed it to `bundle lock` instead.

**Review by rarkins — CHANGES_REQUESTED** [review]
We need to make sure that the `if (output.match(resolveMatchRe)) {` logic isn't executed if it's a lock file maintenance PR. i.e. we don't want it looping and retrying if the initial `bundle lock` fails.

**Comment by bilby91** [comment]
@rarkins Added the check for the loop. Not sure why, but for some reason it seems that when calling `mockExecAll` and passing a second argument with an error, that error is memoized for subsequent tests.

Any ideas ?

**Comment by rarkins** [comment]
@bilby91 sorry, no idea - haven't used it

**Comment by bilby91** [comment]
@rarkins All tests are passing. They might break if the order is changed. Is there someone we could ping to get some insight on the `mockExecAll` ?

**Comment by JamieMagee** [comment]
@bilby91 It's due to the nested `describe` blocks. The current `beforeEach` will run before each nested `describe` block, so the `exec` mocks aren't reset between the tests you wrote.

You've got 2 options:
1. Add a `beforeEach` to your `describe` block for lockfile updates.
2. Remove your `describe` block, and just have your tests run under the top level `describe` block.

Personally, I have a slight preference to number 2, as your tests don't have any specific setup or teardown that they need. So I don't really see the need to group them in the their own `describe` block.

**Comment by rarkins** [comment]
Should the tests be updated before merging?

**Comment by JamieMagee** [comment]
@rarkins Yes. I'll leave it to @bilby91, but I can pick it up if he doesn't have time.

**Comment by bilby91** [comment]
@JamieMagee Thanks for the insight!

I've tried both approaches and none of them seem to be solving the problem. If I make the lock file tests top level and put them before the `Docker` ones, the `Docker` ones will fail. 

Adding a nested beforeEach that calls `jest.resetAllMocks()` doesn't seem to change the result.

**Comment by bilby91** [comment]
@JamieMagee Any clue or other direction to investigate ? 

Thanks for helping!

**Comment by JamieMagee** [comment]
@bilby91 We were also missing `global.repoCache = {};` in `beforeEach`. It wasn't obvious, due to the use of the global object. And we're actively trying to phase out usage of it.

**Comment by bilby91** [comment]
Should be good to merge now ? Thanks @JamieMagee

**Review by JamieMagee — APPROVED** [review]

**Comment by benlangfeld** [comment]
[The documentation](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) still states the following:

> Supported lock files are package-lock.json, yarn.lock, composer.lock and poetry.lock. Others may be added via feature request.

**Pull request merged**
